examples/2_intermediate/mta.ipynb

```diff
@@ -0,0 +1,332 @@
+    "\n",
+    "In our case, \"ticks\" correspond to feed refreshes, and we'll observe this data being updated every 30s."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "f808ff5b-2735-4b58-bd29-054dabdc6620",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import csp\n",
+    "from csp.impl.pushadapter import PushInputAdapter\n",
+    "from csp.impl.wiring import py_push_adapter_def\n",
+    "\n",
+    "import nyct_gtfs\n",
+    "\n",
+    "import os\n",
+    "import time\n",
+    "import threading\n",
+    "from datetime import datetime, timedelta\n",
+    "\n",
+    "API_KEY = os.environ.get('API_KEY')\n",
+    "\n",
+    "\n",
+    "class Event(csp.Struct):\n",
+    "    train: nyct_gtfs.trip.Trip\n",
+    "    update: nyct_gtfs.stop_time_update.StopTimeUpdate\n",
+    "\n",
+    "\n",
+    "class FetchTrainDataAdapter(PushInputAdapter):\n",
+    "    def __init__(self, interval):\n",
+    "        self._interval = interval\n",
+    "        self._thread = None\n",
+    "        self._running = False\n",
+    "\n",
+    "    def start(self, starttime, endtime):\n",
+    "        print(\"FetchTrainDataAdapter::start\")\n",
+    "        self._running = True\n",
+    "        self._thread = threading.Thread(target=self._run)\n",
+    "        self._thread.start()\n",
+    "\n",
+    "    def stop(self):\n",
+    "        print(\"FetchTrainDataAdapter::stop\")\n",
+    "        if self._running:\n",
+    "            self._running = False\n",
+    "            self._thread.join()\n",
+    "\n",
+    "    def _run(self):\n",
+    "        feed = nyct_gtfs.NYCTFeed(\"https://api-endpoint.mta.info/Dataservice/mtagtfsfeeds/nyct%2Fgtfs\", api_key=API_KEY)\n",
+    "\n",
+    "        while self._running:\n",
+    "            print(f\"{datetime.utcnow()}: refreshing MTA feed\")\n",
+    "            feed.refresh()\n",
+    "            # trains will contain all trains underway currently headed to 34St-Penn Station.\n",
+    "            trains = feed.filter_trips(underway=True, headed_for_stop_id=['128', '128N'])\n",
+    "            # tick whenever feed is refreshed\n",
+    "            for train in trains:\n",
+    "                for update in train.stop_time_updates:\n",
+    "                    if update.stop_id in ['128', '128N']:\n",
+    "                        #self.push_tick(Event(train=(train.route_id, train.headsign_text), update=update.arrival))\n",
+    "                        self.push_tick(Event(train=train, update=update))\n",
+    "            time.sleep(self._interval.total_seconds())\n",
+    "\n",
+    "FetchTrainData = py_push_adapter_def(\"FetchTrainData\", FetchTrainDataAdapter, csp.ts[Event], interval=timedelta)\n",
+    "\n",
+    "@csp.graph\n",
+    "def mta_graph():\n",
+    "    print(\"Start of graph building\")\n",
+    "    trains_at_penn = FetchTrainData(timedelta(seconds=30))\n",
+    "    csp.print(\"MTA data\", trains_at_penn)\n",
```

Comment:
is there any other type of calculation you think we can add here to spice it up a bit?

Reply:
Sure, let me think about something.